// @ableton-extensions/cli/dist/run-D3flfXxA.mjs
import { createRequire } from "node:module";
import * as fs from "node:fs";
import * as path from "node:path";
import archiver from "archiver";
import { spawn } from "node:child_process";
//#region src/package.ts
function isInsideDir(parent, child) {
	const rel = path.relative(parent, child);
	return !rel.startsWith("..") && !path.isAbsolute(rel);
}
async function packageExtension(dir, outputPath, include) {
	const extensionDir = path.resolve(dir || process.cwd());
	const manifestPath = path.join(extensionDir, "manifest.json");
	if (!fs.existsSync(manifestPath)) {
		console.error(`No manifest.json found in ${extensionDir}`);
		process.exit(1);
	}
	const manifest = JSON.parse(fs.readFileSync(manifestPath, "utf-8"));
	const entryFile = manifest.entry;
	if (!entryFile) {
		console.error("manifest.json is missing the 'entry' field");
		process.exit(1);
	}
	const entryPath = path.join(extensionDir, entryFile);
	if (!fs.existsSync(entryPath)) {
		console.error(`Entry file not found: ${entryPath}`);
		console.error("Did you build the extension first?");
		process.exit(1);
	}
	for (const inc of include || []) {
		if (path.isAbsolute(inc)) {
			console.error(`Included path must be relative: ${inc}`);
			process.exit(1);
		}
		const resolved = path.resolve(extensionDir, inc);
		if (!isInsideDir(extensionDir, resolved)) {
			console.error(`Included path escapes extension directory: ${inc}`);
			process.exit(1);
		}
		if (!fs.existsSync(resolved)) {
			console.error(`Included path not found: ${inc}`);
			process.exit(1);
		}
	}
	const ablxName = `${(manifest.name || path.basename(extensionDir)).replace(/\s+/gu, "-")}-${manifest.version || "0.0.0"}.ablx`;
	const ablxPath = outputPath || path.join(extensionDir, ablxName);
	const output = fs.createWriteStream(ablxPath);
	const archive = archiver("zip", { zlib: { level: 9 } });
	const done = new Promise((resolve, reject) => {
		output.on("close", resolve);
		archive.on("error", reject);
	});
	archive.pipe(output);
	archive.file(manifestPath, { name: "manifest.json" });
	archive.file(entryPath, { name: entryFile });
	for (const inc of include || []) {
		const resolved = path.resolve(extensionDir, inc);
		if (fs.statSync(resolved).isDirectory()) archive.directory(resolved, inc);
		else archive.file(resolved, { name: inc });
	}
	await archive.finalize();
	await done;
	console.log(ablxPath);
}
//#endregion
//#region src/run.ts
const NODE_MODULE_FILE = "ExtensionHostNodeModule.node";
function resolveExtensionHostDir(livePath) {
	const appPath = livePath.replace(/[\\/]+$/u, "");
	if (!fs.existsSync(appPath)) throw new Error(`Path not found: ${appPath}`);
	if (appPath.endsWith(".node")) return path.dirname(appPath);
	if (appPath.endsWith(".exe")) return path.join(path.dirname(appPath), "ExtensionHost");
	if (appPath.endsWith(".app")) return path.join(appPath, "Contents", "Helpers", "ExtensionHost");
	const winExtHost = path.join(appPath, "Program", "ExtensionHost");
	if (fs.existsSync(path.join(winExtHost, "ExtensionHostNodeModule.node"))) return winExtHost;
	if (fs.statSync(appPath).isDirectory() && fs.existsSync(path.join(appPath, "ExtensionHostNodeModule.node"))) return appPath;
	throw new Error(`Invalid Ableton Live path: ${livePath}\nExpected Ableton Live (.app on macOS, .exe on Windows), a directory containing ExtensionHostNodeModule.node, or a direct path to ExtensionHostNodeModule.node`);
}
function findExtensionHost(livePath) {
	const extensionHostDir = resolveExtensionHostDir(livePath);
	const nodeModule = path.join(extensionHostDir, NODE_MODULE_FILE);
	const nodeExeName = process.platform === "win32" ? "node.exe" : "node";
	const bundledNode = path.join(extensionHostDir, nodeExeName);
	const nodeExecutable = fs.existsSync(bundledNode) ? bundledNode : process.execPath;
	if (!fs.existsSync(nodeModule)) throw new Error(`${NODE_MODULE_FILE} not found at: ${nodeModule}\nMake sure you're using a version of Ableton Live that supports extensions.`);
	return {
		nodeExecutable,
		nodeModule
	};
}
function buildExtensionObject(extensionPath, options) {
	const toForwardSlash = (str) => str.replace(/\\/gu, "/");
	const extensionObject = { path: toForwardSlash(extensionPath) };
	if (options.storageDirectory) extensionObject.storageDirectory = toForwardSlash(path.resolve(options.storageDirectory));
	if (options.tempDirectory) extensionObject.tempDirectory = toForwardSlash(path.resolve(options.tempDirectory));
	return extensionObject;
}
function runExtension(extensionDir, options) {
	const { nodeExecutable, nodeModule } = findExtensionHost(options.live);
	const extensionPath = path.resolve(extensionDir || process.cwd());
	const pkgPath = path.join(extensionPath, "package.json");
	if (!fs.existsSync(pkgPath)) {
		console.error(`No package.json found at: ${extensionPath}`);
		process.exit(1);
	}
	console.log("Starting Extension Host...");
	console.log(`  Extension: ${extensionPath}`);
	console.log(`  Live: ${options.live}`);
	console.log();
	const extensionObject = buildExtensionObject(extensionPath, options);
	if (options.inspect) {
		const require = createRequire(import.meta.url);
		globalThis.require = require;
		require(nodeModule).initialize({ extensions: [extensionObject] });
		return;
	}
	spawn(nodeExecutable, ["-e", `require('${nodeModule.replace(/\\/gu, "/")}').initialize(${JSON.stringify({ extensions: [extensionObject] })});`], {
		cwd: process.cwd(),
		stdio: "inherit"
	}).on("exit", (code) => {
		process.exit(code || 0);
	});
}
//#endregion
export { packageExtension as n, runExtension as t };
