// @ableton-extensions/cli/dist/run-pQFYI5zn.cjs
//#region \0rolldown/runtime.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __copyProps = (to, from, except, desc) => {
	if (from && typeof from === "object" || typeof from === "function") for (var keys = __getOwnPropNames(from), i = 0, n = keys.length, key; i < n; i++) {
		key = keys[i];
		if (!__hasOwnProp.call(to, key) && key !== except) __defProp(to, key, {
			get: ((k) => from[k]).bind(null, key),
			enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable
		});
	}
	return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", {
	value: mod,
	enumerable: true
}) : target, mod));
//#endregion
let node_fs = require("node:fs");
node_fs = __toESM(node_fs, 1);
let node_path = require("node:path");
node_path = __toESM(node_path, 1);
let archiver = require("archiver");
archiver = __toESM(archiver, 1);
let node_module = require("node:module");
let node_child_process = require("node:child_process");
//#region src/package.ts
function isInsideDir(parent, child) {
	const rel = node_path.relative(parent, child);
	return !rel.startsWith("..") && !node_path.isAbsolute(rel);
}
async function packageExtension(dir, outputPath, include) {
	const extensionDir = node_path.resolve(dir || process.cwd());
	const manifestPath = node_path.join(extensionDir, "manifest.json");
	if (!node_fs.existsSync(manifestPath)) {
		console.error(`No manifest.json found in ${extensionDir}`);
		process.exit(1);
	}
	const manifest = JSON.parse(node_fs.readFileSync(manifestPath, "utf-8"));
	const entryFile = manifest.entry;
	if (!entryFile) {
		console.error("manifest.json is missing the 'entry' field");
		process.exit(1);
	}
	const entryPath = node_path.join(extensionDir, entryFile);
	if (!node_fs.existsSync(entryPath)) {
		console.error(`Entry file not found: ${entryPath}`);
		console.error("Did you build the extension first?");
		process.exit(1);
	}
	for (const inc of include || []) {
		if (node_path.isAbsolute(inc)) {
			console.error(`Included path must be relative: ${inc}`);
			process.exit(1);
		}
		const resolved = node_path.resolve(extensionDir, inc);
		if (!isInsideDir(extensionDir, resolved)) {
			console.error(`Included path escapes extension directory: ${inc}`);
			process.exit(1);
		}
		if (!node_fs.existsSync(resolved)) {
			console.error(`Included path not found: ${inc}`);
			process.exit(1);
		}
	}
	const ablxName = `${(manifest.name || node_path.basename(extensionDir)).replace(/\s+/gu, "-")}-${manifest.version || "0.0.0"}.ablx`;
	const ablxPath = outputPath || node_path.join(extensionDir, ablxName);
	const output = node_fs.createWriteStream(ablxPath);
	const archive = (0, archiver.default)("zip", { zlib: { level: 9 } });
	const done = new Promise((resolve, reject) => {
		output.on("close", resolve);
		archive.on("error", reject);
	});
	archive.pipe(output);
	archive.file(manifestPath, { name: "manifest.json" });
	archive.file(entryPath, { name: entryFile });
	for (const inc of include || []) {
		const resolved = node_path.resolve(extensionDir, inc);
		if (node_fs.statSync(resolved).isDirectory()) archive.directory(resolved, inc);
		else archive.file(resolved, { name: inc });
	}
	await archive.finalize();
	await done;
	console.log(ablxPath);
}
//#endregion
//#region src/run.ts
const NODE_MODULE_FILE = "ExtensionHostNodeModule.node";
function resolveExtensionHostDir(livePath) {
	const appPath = livePath.replace(/[\\/]+$/u, "");
	if (!node_fs.existsSync(appPath)) throw new Error(`Path not found: ${appPath}`);
	if (appPath.endsWith(".node")) return node_path.dirname(appPath);
	if (appPath.endsWith(".exe")) return node_path.join(node_path.dirname(appPath), "ExtensionHost");
	if (appPath.endsWith(".app")) return node_path.join(appPath, "Contents", "Helpers", "ExtensionHost");
	const winExtHost = node_path.join(appPath, "Program", "ExtensionHost");
	if (node_fs.existsSync(node_path.join(winExtHost, "ExtensionHostNodeModule.node"))) return winExtHost;
	if (node_fs.statSync(appPath).isDirectory() && node_fs.existsSync(node_path.join(appPath, "ExtensionHostNodeModule.node"))) return appPath;
	throw new Error(`Invalid Ableton Live path: ${livePath}\nExpected Ableton Live (.app on macOS, .exe on Windows), a directory containing ExtensionHostNodeModule.node, or a direct path to ExtensionHostNodeModule.node`);
}
function findExtensionHost(livePath) {
	const extensionHostDir = resolveExtensionHostDir(livePath);
	const nodeModule = node_path.join(extensionHostDir, NODE_MODULE_FILE);
	const nodeExeName = process.platform === "win32" ? "node.exe" : "node";
	const bundledNode = node_path.join(extensionHostDir, nodeExeName);
	const nodeExecutable = node_fs.existsSync(bundledNode) ? bundledNode : process.execPath;
	if (!node_fs.existsSync(nodeModule)) throw new Error(`${NODE_MODULE_FILE} not found at: ${nodeModule}\nMake sure you're using a version of Ableton Live that supports extensions.`);
	return {
		nodeExecutable,
		nodeModule
	};
}
function buildExtensionObject(extensionPath, options) {
	const toForwardSlash = (str) => str.replace(/\\/gu, "/");
	const extensionObject = { path: toForwardSlash(extensionPath) };
	if (options.storageDirectory) extensionObject.storageDirectory = toForwardSlash(node_path.resolve(options.storageDirectory));
	if (options.tempDirectory) extensionObject.tempDirectory = toForwardSlash(node_path.resolve(options.tempDirectory));
	return extensionObject;
}
function runExtension(extensionDir, options) {
	const { nodeExecutable, nodeModule } = findExtensionHost(options.live);
	const extensionPath = node_path.resolve(extensionDir || process.cwd());
	const pkgPath = node_path.join(extensionPath, "package.json");
	if (!node_fs.existsSync(pkgPath)) {
		console.error(`No package.json found at: ${extensionPath}`);
		process.exit(1);
	}
	console.log("Starting Extension Host...");
	console.log(`  Extension: ${extensionPath}`);
	console.log(`  Live: ${options.live}`);
	console.log();
	const extensionObject = buildExtensionObject(extensionPath, options);
	if (options.inspect) {
		const require = (0, node_module.createRequire)(require("url").pathToFileURL(__filename).href);
		globalThis.require = require;
		require(nodeModule).initialize({ extensions: [extensionObject] });
		return;
	}
	(0, node_child_process.spawn)(nodeExecutable, ["-e", `require('${nodeModule.replace(/\\/gu, "/")}').initialize(${JSON.stringify({ extensions: [extensionObject] })});`], {
		cwd: process.cwd(),
		stdio: "inherit"
	}).on("exit", (code) => {
		process.exit(code || 0);
	});
}
//#endregion
Object.defineProperty(exports, "__toESM", {
	enumerable: true,
	get: function() {
		return __toESM;
	}
});
Object.defineProperty(exports, "packageExtension", {
	enumerable: true,
	get: function() {
		return packageExtension;
	}
});
Object.defineProperty(exports, "runExtension", {
	enumerable: true,
	get: function() {
		return runExtension;
	}
});
